### ISSUE TITLE
When running fdp using ioengine as io_uring_cmd on nvme device with lba format 4K+64, "only extended logical block can be supported" error occurred.

### ISSUE THREAD

**Issue by ChaeLinYeo** [axboe/fio#1613]
**Please acknowledge the following before creating a ticket**

- [Yes] I have read the GitHub issues section of [REPORTING-BUGS](../blob/master/REPORTING-BUGS).

**Description of the bug:**
When running fdp using ioengine as io_uring_cmd on nvme device with lba format 4K+64, "only extended logical block can be supported" error occurred.

**Environment**: 
Linux fadu 6.2.10 #2 SMP PREEMPT_DYNAMIC Mon Apr 17 10:27:43 KST 2023 x86_64 x86_64 x86_64 GNU/Linux

**fio version**: 
3.35

**Reproduction steps**
nvme list:  
Node                  Generic               Namespace  Usage                      Format           
--------------------- --------------------- ---------- -------------------------- ----------------
/dev/nvme0n2          /dev/ng0n2            0x2          1.09  TB /   1.82  TB      4 KiB + 64 B   
/dev/nvme0n1          /dev/ng0n1            0x1          1.08  TB /   1.82  TB      4 KiB +  0 B   

fio workload cmd:
/usr/local/bin/fio --name=global --output-format=json+ --ioengine=io_uring_cmd --time_based=1 --runtime=10 --cmd_type=nvme --iodepth=32 --bs=4k --fdp=1 --filename=/dev/ng0n1 --name=write_heavy --rw=randrw --rwmixwrite=90 --fdp_pli=0,1,2,3 --offset=0% --size=30% --name=write_mid --rw=randrw --rwmixwrite=30 --fdp_pli=4,5 --offset=30% --size=30% --name=write_light --rw=randrw --rwmixwrite=10 --fdp_pli=6 --offset=60% --size=30%

Even fdp fio workload, which performs the simplest read, does not work.
/usr/local/bin/fio --name=global --ioengine=io_uring_cmd --time_based=1 --runtime=10 --cmd_type=nvme --fdp=1 --filename=/dev/ng0n1 --name=job0 --rw=read
  
error msg: only extended logical block can be supported

https://github.com/axboe/fio/releases
I tested it using fio-3.35 from the release link above and found that the conditional statement at line 164 of the /engines/nvme.c file caused the error.
conditional statement: if (*ms && !((ns.flbas >> 4) & 0x1))
When executing the nvme id-ns command, ms of my device namespace with lba format 4K+64 is 64 and flbas is 0x1. Therefore, it is proved that the conditinal statement is true and it caused the error.

It seems that FDP should be supported regardless of whether bit 4 of flbas is 1 or 0. Why was this exception handling added? If it's not fio's problem, is it the nvme device I'm using?

**Comment by ankit-sam** [comment]
Hi @ChaeLinYeo please use the fio master branch. fio-3.35 does not support separate metadata buffer and protection information. We recently added that support. For metadata buffer size you will need to specify option `md_per_io_size` as per the block size.
Ex - For 4K block size, `md_per_io_size` should be 64. For 16K block size, `md_per_io_size` should be 256.

**Comment by fadu-jenna** [comment]
@ankit-sam , Thank you for your reply. Additionally, as a result of testing other versions, in fio-3.34, FDP workload worked without errors as there was no conditional statement exception handling for ms and flbas in /engines/nvme.c. However, if support of separate metadata buffer and protection information was recently added to the master branch, will running FDP workloads using fio-3.34 cause unexpected problems?

**Comment by ankit-sam** [comment]
Hi @ChaeLinYeo
With fio-3.34 there was no support for metadata buffer. So the nvme passthrough commands was formed without it. For fio-3.35 we added support for extended logical blocks (metadata at the end of logical block) such as 520 (512 + 8), 4160 (4096 + 64) etc, but it still did not support separate metadata buffers. We added checks during initialization itself to prevent any error during read/write.

I tried with fio-3.34 on qemu as well as actual hardware, with LBA format (4K + 64B separate metadata buffer).
With qemu even if you don't pass metadata buffer, the NVMe command succeeds, its most probably a bug, should fail.
With the actual device I see failures, and dmesg shows I/O error with (sct 0x0 / sc 0x4). As per NVMe base specification its
Data Transfer Error: Transferring the data or metadata associated with a command had an error

Anyway, I will suggest you to use the latest fio master branch.
